nixos/modules/services/networking/frp-client.nix

```diff
@@ -0,0 +1,82 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.frp-client;
+  settingsFormat = pkgs.formats.ini { };
+in
+{
+  options = {
+    services.frp-client = {
+      enable = mkEnableOption (lib.mdDoc "frp client configuration");
+
+      package = mkPackageOptionMD pkgs "frp" { };
+
+      settings = mkOption {
+        type = settingsFormat.type;
+        default = { };
+        description = lib.mdDoc ''
+          frp client configuration, for configuration options see the example on [github](https://github.com/fatedier/frp/blob/dev/conf/frpc_full.ini)
+        '';
+        example = literalExpression ''
+          {
+            common = {
+              server_addr = "x.x.x.x";
+              server_port = 7000;
+            };
+          }
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment = {
+      etc."frpc.ini".source = settingsFormat.generate "frpc.ini" cfg.settings;
+    };
+    systemd.services = {
+      frp-client = {
+        wants = [ "network.target" ];
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        description = "frp client";
+        serviceConfig = {
+          DynamicUser = true;
```

Comment:
Some hardending options, such as `ProtectSystem = "strict"`, are implied by `DynamicUser = true`.

FYI, `systemd-analyze security your-service` can be used to check the final result.